### PR TITLE
Implement element access for inline array types

### PR DESCRIPTION
**Pull request by AlekseyTs**
https://github.com/dotnet/csharplang/blob/main/proposals/inline-arrays.md#element-access

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs [hunk 1]
```diff
@@ -167,6 +167,138 @@ private BoundExpression VisitIndexerAccess(BoundIndexerAccess node, bool isLeftO
...
+                                               _factory.Call(_factory.Call(null, createSpan, possiblyRefCapturedReceiver, _factory.Literal(length)),
+                                                             getItemOrSliceHelper, startExpr, rangeSizeExpr));
+                }
+            }
+
+            if (receiverStore is not null)
+            {
+                result = _factory.Sequence(ImmutableArray.Create(((BoundLocal)rewrittenReceiver).LocalSymbol),
+                                           ImmutableArray.Create((BoundExpression)receiverStore),
+                                           result);// RefAccessMustMakeCopy(result));
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_IndexerAccess.cs:279]: > `// RefAccessMustMakeCopy(result));`

Forgotten commented code?

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs [hunk 2]
```diff
@@ -412,5 +412,10 @@ internal override IEnumerable<(MethodSymbol Body, MethodSymbol Implemented)> Syn
                 }
             }
         }
+
+        internal override bool HasInlineArrayAttribute(out int length)
+        {
+            return _underlyingType.HasInlineArrayAttribute(out length);
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs:418]: Do I understand correctly that RetargetingNamedTypeSymbol is used for type forwarding? Should we have a test verifying that type forwarding works with inline array struct definitions?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs:418]: > Do I understand correctly that RetargetingNamedTypeSymbol is used for type forwarding?

No. Retargeting symbols have no primary relation with type forwarding.

> Should we have a test verifying that type forwarding works with inline array struct definitions?

Given trivial implementation, the only reason to have a test is completeness, I think.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -184,7 +184,7 @@ public VisitArgumentResult(VisitResult visitResult, Optional<LocalState> stateFo
         /// Invalid type, used only to catch Visit methods that do not set
         /// _result.Type. See VisitExpressionWithoutStackGuard.
         /// </summary>
-        private static readonly TypeWithState _invalidType = TypeWithState.Create(ErrorTypeSymbol.UnknownResultType, NullableFlowState.NotNull);
+        private static readonly TypeWithState _invalidType = TypeWithState.Create(new UnsupportedMetadataTypeSymbol(), NullableFlowState.NotNull);
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:187]: > Why this change?

The ```_invalidType``` is sentinel. A sentinel must have unique value. The ```ErrorTypeSymbol.UnknownResultType``` is a shared instance, anyone is free to use it, given a reason. That usage should not be confused with the sentinel value.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 4]
```diff
@@ -1118,6 +1118,17 @@ protected sealed override void DecodeWellKnownAttributeImpl(ref DecodeWellKnownA
                 // CS1608: The Required attribute is not permitted on C# types
                 diagnostics.Add(ErrorCode.ERR_CantUseRequiredAttribute, arguments.AttributeSyntaxOpt.Name.Location);
             }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.InlineArrayAttribute))
+            {
+                if (arguments.GetOrCreateData<TypeWellKnownAttributeData>().InlineArrayLength == 0)
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1123]: > Why don't we use ```arguments.HasDecodedData``` to detect that attribute hasn't been decoded yet?

Because that property doesn't return information specific to this attribute.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 5]
```diff
@@ -2466,5 +2466,34 @@ ITypeSymbol ITypeSymbolInternal.GetITypeSymbol()
...
+                {
+                    if (member is FieldSymbol { IsStatic: false } field)
+                    {
+                        if (field.RefKind != RefKind.None || elementType.HasType)
+                        {
+                            return default;
+                        }
+                        else
+                        {
+                            elementType = field.TypeWithAnnotations;
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs:2490]: > It looks like ```elementType``` will be the type of the last field member of the inline array struct. Shouldn't it be the first field member?

How so? Could you elaborate?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 6]
```diff
@@ -7944,11 +7946,133 @@ private BoundExpression BindElementOrIndexerAccess(ExpressionSyntax node, BoundE
                 // able to get more semantic analysis of the indexing operation. We do not
                 // want to report cascading errors.
 
-                BoundExpression result = BindElementAccessCore(node, expr, analyzedArguments, BindingDiagnosticBag.Discarded);
-                return result;
+                diagnostics = BindingDiagnosticBag.Discarded;
+            }
+
+            if (expr.Type.HasInlineArrayAttribute(out _) == true && analyzedArguments.Arguments.Count == 1 && expr.Type.TryGetInlineArrayElementType() is { HasType: true } elementType &&
+                tryImplicitConversionToFixedBufferIndex(node, analyzedArguments.Arguments[0], diagnostics, out WellKnownType indexOrRangeWellknownType) is { } convertedIndex)
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7953]: Prefer keeping explicit ```Count``` check for readability. Complex patterns often hard to follow.

File: src/Compilers/CSharp/Portable/BoundTree/BoundInlineArrayAccess.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class BoundInlineArrayAccess
+    {
+        private partial void Validate()
+        {
+#if DEBUG
+            Debug.Assert(!IsValue || GetItemOrSliceHelper == WellKnownMember.System_ReadOnlySpan_T__get_Item);
+#pragma warning disable format
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/BoundTree/BoundInlineArrayAccess.cs:17]: > I think if you use parenthesis around RHS of is, you can accomplish nice formatting without disabling the warning.

The suggested formatting might be nice, but not the formatting I want to use.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingNamedTypeSymbol.cs:418]: Added ```InlineArrayType_18_Retargeting``` test

File: src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs [hunk 8]
```diff
@@ -194,6 +194,7 @@ static AttributeDescription()
         private static readonly byte[][] s_signatures_HasThis_Void_String_Only = { s_signature_HasThis_Void_String };
         private static readonly byte[][] s_signatures_HasThis_Void_Type_Only = { s_signature_HasThis_Void_Type };
         private static readonly byte[][] s_signatures_HasThis_Void_Boolean_Only = { s_signature_HasThis_Void_Boolean };
+        private static readonly byte[][] s_signature_HasThis_Void_Int32_Only = { s_signature_HasThis_Void_Int32 };
```
Comment (cston) [inline src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:197]: >s_signature [](http://example.com/codeflow?start=41&length=11)

Consider naming `s_signatures_...` (plural) for consistency.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 9]
```diff
@@ -726,6 +726,28 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
...
+                        }
+
+                        var getItemOrSliceHelper = (MethodSymbol)Compilation.GetWellKnownTypeMember(elementAccess.GetItemOrSliceHelper);
+
+                        if (getItemOrSliceHelper is null)
+                        {
+                            return true;
+                        }
+
+                        getItemOrSliceHelper = getItemOrSliceHelper.AsMember(getItemOrSliceHelper.ContainingType.Construct(ImmutableArray.Create(elementAccess.Expression.Type.TryGetInlineArrayElementType())));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:746]: >TryGetInlineArrayElementType() [](http://example.com/codeflow?start=175&length=30)

Can this return `TypeWithAnnotations { HasType: false }`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 10]
```diff
@@ -7944,11 +7946,133 @@ private BoundExpression BindElementOrIndexerAccess(ExpressionSyntax node, BoundE
                 // able to get more semantic analysis of the indexing operation. We do not
                 // want to report cascading errors.
 
-                BoundExpression result = BindElementAccessCore(node, expr, analyzedArguments, BindingDiagnosticBag.Discarded);
-                return result;
+                diagnostics = BindingDiagnosticBag.Discarded;
+            }
+
+            if (expr.Type.HasInlineArrayAttribute(out _) == true && analyzedArguments.Arguments.Count == 1 && expr.Type.TryGetInlineArrayElementType() is { HasType: true } elementType &&
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7952]: > == true [](http://example.com/codeflow?start=56&length=8)

Can be removed.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 11]
```diff
@@ -7944,11 +7946,133 @@ private BoundExpression BindElementOrIndexerAccess(ExpressionSyntax node, BoundE
...
+                tryImplicitConversionToFixedBufferIndex(node, analyzedArguments.Arguments[0], diagnostics, out WellKnownType indexOrRangeWellknownType) is { } convertedIndex)
+            {
+                return bindFixedBufferElementAccess(node, expr, analyzedArguments, convertedIndex, indexOrRangeWellknownType, elementType, diagnostics);
             }
 
             return BindElementAccessCore(node, expr, analyzedArguments, diagnostics);
+
+            BoundExpression tryImplicitConversionToFixedBufferIndex(ExpressionSyntax node, BoundExpression index, BindingDiagnosticBag diagnostics, out WellKnownType indexOrRangeWellknownType)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7960]: >FixedBuffer [](http://example.com/codeflow?start=51&length=11)

Rename to `tryImplicitConversionToInlineArrayIndex`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 12]
```diff
@@ -7944,11 +7946,133 @@ private BoundExpression BindElementOrIndexerAccess(ExpressionSyntax node, BoundE
...
+                    else
+                    {
+                        indexOrRangeWellknownType = WellKnownType.System_Index;
+                    }
+                }
+
+                return convertedIndex;
+            }
+
+            BoundExpression bindFixedBufferElementAccess(ExpressionSyntax node, BoundExpression expr, AnalyzedArguments analyzedArguments, BoundExpression convertedIndex, WellKnownType indexOrRangeWellknownType, TypeWithAnnotations elementType, BindingDiagnosticBag diagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7986]: >bindFixedBufferElementAccess [](http://example.com/codeflow?start=28&length=28)

Rename to `bindInlineArrayElementAccess`?

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 9 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:746]: > Can this return ```TypeWithAnnotations { HasType: false }```?

Not for receiver of ```BoundInlineArrayAccess```.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 10 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7952]:
```suggestion
            if (expr.Type.HasInlineArrayAttribute(out _) && analyzedArguments.Arguments.Count == 1 && expr.Type.TryGetInlineArrayElementType() is { HasType: true } elementType &&
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 11 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7960]:
```suggestion
            BoundExpression tryImplicitConversionToInlineArrayIndex(ExpressionSyntax node, BoundExpression index, BindingDiagnosticBag diagnostics, out WellKnownType indexOrRangeWellknownType)
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7953]:
```suggestion
                tryImplicitConversionToInlineArrayIndex(node, analyzedArguments.Arguments[0], diagnostics, out WellKnownType indexOrRangeWellknownType) is { } convertedIndex)
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 13]
```diff
@@ -7944,11 +7946,133 @@ private BoundExpression BindElementOrIndexerAccess(ExpressionSyntax node, BoundE
...
 
-                BoundExpression result = BindElementAccessCore(node, expr, analyzedArguments, BindingDiagnosticBag.Discarded);
-                return result;
+                diagnostics = BindingDiagnosticBag.Discarded;
+            }
+
+            if (expr.Type.HasInlineArrayAttribute(out _) == true && analyzedArguments.Arguments.Count == 1 && expr.Type.TryGetInlineArrayElementType() is { HasType: true } elementType &&
+                tryImplicitConversionToFixedBufferIndex(node, analyzedArguments.Arguments[0], diagnostics, out WellKnownType indexOrRangeWellknownType) is { } convertedIndex)
+            {
+                return bindFixedBufferElementAccess(node, expr, analyzedArguments, convertedIndex, indexOrRangeWellknownType, elementType, diagnostics);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7955]:
```suggestion
                return bindInlineArrayElementAccess(node, expr, analyzedArguments, convertedIndex, indexOrRangeWellknownType, elementType, diagnostics);
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 12 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7986]:
```suggestion
            BoundExpression bindInlineArrayElementAccess(ExpressionSyntax node, BoundExpression expr, AnalyzedArguments analyzedArguments, BoundExpression convertedIndex, WellKnownType indexOrRangeWellknownType, TypeWithAnnotations elementType, BindingDiagnosticBag diagnostics)
```

File: src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs (hunk 8 above)
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:197]:
```suggestion
        private static readonly byte[][] s_signatures_HasThis_Void_Int32_Only = { s_signature_HasThis_Void_Int32 };
```

File: src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs [hunk 14]
```diff
@@ -382,7 +380,7 @@ static AttributeDescription()
         internal static readonly AttributeDescription SerializableAttribute = new AttributeDescription("System", "SerializableAttribute", s_signatures_HasThis_Void_Only);
         internal static readonly AttributeDescription NonSerializedAttribute = new AttributeDescription("System", "NonSerializedAttribute", s_signatures_HasThis_Void_Only);
         internal static readonly AttributeDescription StructLayoutAttribute = new AttributeDescription("System.Runtime.InteropServices", "StructLayoutAttribute", s_signaturesOfStructLayoutAttribute);
-        internal static readonly AttributeDescription FieldOffsetAttribute = new AttributeDescription("System.Runtime.InteropServices", "FieldOffsetAttribute", s_signaturesOfFieldOffsetAttribute);
+        internal static readonly AttributeDescription FieldOffsetAttribute = new AttributeDescription("System.Runtime.InteropServices", "FieldOffsetAttribute", s_signature_HasThis_Void_Int32_Only);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:383]:
```suggestion
        internal static readonly AttributeDescription FieldOffsetAttribute = new AttributeDescription("System.Runtime.InteropServices", "FieldOffsetAttribute", s_signatures_HasThis_Void_Int32_Only);
```

File: src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs [hunk 15]
```diff
@@ -419,7 +417,7 @@ static AttributeDescription()
         internal static readonly AttributeDescription ClassInterfaceAttribute = new AttributeDescription("System.Runtime.InteropServices", "ClassInterfaceAttribute", s_signaturesOfClassInterfaceAttribute);
         internal static readonly AttributeDescription ComSourceInterfacesAttribute = new AttributeDescription("System.Runtime.InteropServices", "ComSourceInterfacesAttribute", s_signaturesOfComSourceInterfacesAttribute);
         internal static readonly AttributeDescription ComVisibleAttribute = new AttributeDescription("System.Runtime.InteropServices", "ComVisibleAttribute", s_signatures_HasThis_Void_Boolean_Only);
-        internal static readonly AttributeDescription DispIdAttribute = new AttributeDescription("System.Runtime.InteropServices", "DispIdAttribute", new byte[][] { s_signature_HasThis_Void_Int32 });
+        internal static readonly AttributeDescription DispIdAttribute = new AttributeDescription("System.Runtime.InteropServices", "DispIdAttribute", s_signature_HasThis_Void_Int32_Only);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:420]:
```suggestion
        internal static readonly AttributeDescription DispIdAttribute = new AttributeDescription("System.Runtime.InteropServices", "DispIdAttribute", s_signatures_HasThis_Void_Int32_Only);
```

File: src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs [hunk 16]
```diff
@@ -453,7 +451,7 @@ static AttributeDescription()
         internal static readonly AttributeDescription ComEventInterfaceAttribute = new AttributeDescription("System.Runtime.InteropServices", "ComEventInterfaceAttribute", s_signaturesOfComEventInterfaceAttribute);
         internal static readonly AttributeDescription BestFitMappingAttribute = new AttributeDescription("System.Runtime.InteropServices", "BestFitMappingAttribute", s_signatures_HasThis_Void_Boolean_Only);
         internal static readonly AttributeDescription FlagsAttribute = new AttributeDescription("System", "FlagsAttribute", s_signatures_HasThis_Void_Only);
-        internal static readonly AttributeDescription LCIDConversionAttribute = new AttributeDescription("System.Runtime.InteropServices", "LCIDConversionAttribute", s_signaturesOfLCIDConversionAttribute);
+        internal static readonly AttributeDescription LCIDConversionAttribute = new AttributeDescription("System.Runtime.InteropServices", "LCIDConversionAttribute", s_signature_HasThis_Void_Int32_Only);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:454]:
```suggestion
        internal static readonly AttributeDescription LCIDConversionAttribute = new AttributeDescription("System.Runtime.InteropServices", "LCIDConversionAttribute", s_signatures_HasThis_Void_Int32_Only);
```

File: src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs [hunk 17]
```diff
@@ -470,7 +468,7 @@ static AttributeDescription()
         internal static readonly AttributeDescription SkipLocalsInitAttribute = new AttributeDescription("System.Runtime.CompilerServices", "SkipLocalsInitAttribute", s_signatures_HasThis_Void_Only);
         internal static readonly AttributeDescription NativeIntegerAttribute = new AttributeDescription("System.Runtime.CompilerServices", "NativeIntegerAttribute", s_signaturesOfNativeIntegerAttribute);
         internal static readonly AttributeDescription ScopedRefAttribute = new AttributeDescription("System.Runtime.CompilerServices", "ScopedRefAttribute", s_signatures_HasThis_Void_Only);
-        internal static readonly AttributeDescription RefSafetyRulesAttribute = new AttributeDescription("System.Runtime.CompilerServices", "RefSafetyRulesAttribute", s_signaturesOfRefSafetyRulesAttribute);
+        internal static readonly AttributeDescription RefSafetyRulesAttribute = new AttributeDescription("System.Runtime.CompilerServices", "RefSafetyRulesAttribute", s_signature_HasThis_Void_Int32_Only);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:471]:
```suggestion
        internal static readonly AttributeDescription RefSafetyRulesAttribute = new AttributeDescription("System.Runtime.CompilerServices", "RefSafetyRulesAttribute", s_signatures_HasThis_Void_Int32_Only);
```

File: src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs [hunk 18]
```diff
@@ -479,5 +477,6 @@ static AttributeDescription()
         internal static readonly AttributeDescription SetsRequiredMembersAttribute = new AttributeDescription("System.Diagnostics.CodeAnalysis", "SetsRequiredMembersAttribute", s_signatures_HasThis_Void_Only);
         internal static readonly AttributeDescription CompilerFeatureRequiredAttribute = new AttributeDescription("System.Runtime.CompilerServices", "CompilerFeatureRequiredAttribute", s_signatures_HasThis_Void_String_Only);
         internal static readonly AttributeDescription UnscopedRefAttribute = new AttributeDescription("System.Diagnostics.CodeAnalysis", "UnscopedRefAttribute", s_signatures_HasThis_Void_Only);
+        internal static readonly AttributeDescription InlineArrayAttribute = new AttributeDescription("System.Runtime.CompilerServices", "InlineArrayAttribute", s_signature_HasThis_Void_Int32_Only);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/Attributes/AttributeDescription.cs:480]:
```suggestion
        internal static readonly AttributeDescription InlineArrayAttribute = new AttributeDescription("System.Runtime.CompilerServices", "InlineArrayAttribute", s_signatures_HasThis_Void_Int32_Only);
```

**Comment by AlekseyTs** [comment]
@cston, @jjonescz Please review

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 19]
```diff
@@ -1512,6 +1512,61 @@ protected virtual void AssignImpl(BoundNode node, BoundExpression value, bool is
...
+                                    if (isRef && param.RefKind == RefKind.Out)
+                                    {
+                                        LeaveParameter(param, elementAccess.Expression.Syntax, paramExpr.Syntax.Location);
+                                    }
+
+                                    if (written) NoteWrite(paramExpr, null, read);
+                                    break;
+                                }
+
+                            case BoundKind.ThisReference:
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1558]: >ThisReference [](http://example.com/codeflow?start=43&length=13)

Do we need to handle `BaseReference` as well?

File: src/Compilers/CSharp/Portable/Symbols/SubstitutedMethodSymbol.cs [hunk 20]
```diff
@@ -360,16 +360,24 @@ internal override int CalculateLocalSyntaxOffset(int localPosition, SyntaxTree l
...
+            int containingHashCode;
 
-            // If the containing type of the original definition is the same as our containing type
-            // it's possible that we will compare equal to the original definition under certain conditions 
-            // (e.g, ignoring nullability) and want to retain the same hashcode. As such, consider only
-            // the original definition for the hashcode when we know equality is possible
-            var containingHashCode = _containingType.GetHashCode();
-            if (containingHashCode == this.OriginalDefinition.ContainingType.GetHashCode() &&
-                wasConstructedForAnnotations(this))
+            if (OriginalDefinition is SynthesizedGlobalMethodSymbol global)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/SubstitutedMethodSymbol.cs:365]: >if (OriginalDefinition is SynthesizedGlobalMethodSymbol global) [](http://example.com/codeflow?start=12&length=63)

Why wasn't the `else` case appropriate for `SynthesizedGlobalMethodSymbol`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs (hunk 19 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1558]: > Do we need to handle ```BaseReference``` as well?

I do not think so. It looks like this function doesn't track assignements to the base. If you have a specific scenario in mind, feel free to share.

File: src/Compilers/CSharp/Portable/Symbols/SubstitutedMethodSymbol.cs (hunk 20 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/SubstitutedMethodSymbol.cs:365]: > Why wasn't the else case appropriate for ```SynthesizedGlobalMethodSymbol```?

Containing type for ```SynthesizedGlobalMethodSymbol``` is null.

**Comment by cston** [comment]
>         public void InlineArrayType_07_WrongLength()

Duplicate of `InlineArrayType_05_WrongLength`?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:211 in 4ee91b2. [](commit_id = 4ee91b28e800a30b6ea188dea3eaac19bd752149, deletion_comment = False)

**Comment by cston** [comment]
>     static int M4(in int x, Buffer10<int> y)

What is this parameter testing?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:2682 in 4ee91b2. [](commit_id = 4ee91b28e800a30b6ea188dea3eaac19bd752149, deletion_comment = False)

**Comment by cston** [comment]
>     }

Consider testing `ref struct Buffer10<T> { T element; }` and ensure that uses of the type cannot be boxed or used as a field other than in a `ref struct`.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:6194 in 4ee91b2. [](commit_id = 4ee91b28e800a30b6ea188dea3eaac19bd752149, deletion_comment = False)

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
>     static int M4(in int x, Buffer10<int> y)

> What is this parameter testing?

We are testing that we can pass a value returning element access as an `in` parameter and that reuse of the same temp (```Buffer10<int> V_0```) during emit happens and doesn't cause problems.

---
In reply to: [1515489870](https://github.com/dotnet/roslyn/pull/67830#issuecomment-1515489870) [](http://example.com/codeflow?ancestors=1515489870,1515489870)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:2682 in 4ee91b2. [](commit_id = 4ee91b28e800a30b6ea188dea3eaac19bd752149, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     static int M4(in int x, Buffer10<int> y)

We are not passing a ref to an item in a temp.

---
In reply to: [1515622981](https://github.com/dotnet/roslyn/pull/67830#issuecomment-1515622981) [](http://example.com/codeflow?ancestors=1515622981,1515489870,1515489870)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:2682 in 4ee91b2. [](commit_id = 4ee91b28e800a30b6ea188dea3eaac19bd752149, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     }

> Consider testing ```ref struct Buffer10<T> { T element; }``` and ensure that uses of the type cannot be boxed or used as a field other than in a ref struct.

I do not find this scenario particularly interesting in context of the inline array feature. The feature doesn't depend on ref-ness of the struct and doesn't affect this property of the struct. I prefer to not add tests for a scenario that doesn't have any merit other than the fact that the scenario exists. 


---
In reply to: [1515509830](https://github.com/dotnet/roslyn/pull/67830#issuecomment-1515509830) [](http://example.com/codeflow?ancestors=1515509830,1515509830)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:6194 in 4ee91b2. [](commit_id = 4ee91b28e800a30b6ea188dea3eaac19bd752149, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs [hunk 21]
```diff
@@ -0,0 +1,6195 @@
...
+                Assert.False(buffer.HasInlineArrayAttribute(out int length));
+                Assert.Equal(0, length);
+            }
+        }
+
+        [Fact]
+        public void InlineArrayType_07_WrongLength()
+        {
+            var src = @"
+[System.Runtime.CompilerServices.InlineArray(0)]
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:214]:
```suggestion
[System.Runtime.CompilerServices.InlineArray(-2)]
```

**Comment by AlekseyTs** [comment]
@jjonescz Please review, I think I addressed your feedback.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1123]: I guess this is fine, but seems that using `HasDecodedData` here would be useful. We would not need two special values (0 and -1) for `InlineArrayLength` (and mapping between them below) - the `if` would just check for `!HasDecodedData`. We could also have `VerifySealed(expected: true);` in getter of `TypeWellKnownAttributeData.InlineArrayLength`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs [hunk 22]
```diff
@@ -0,0 +1,6195 @@
...
+        public InlineArrayAttribute (int length)
+        {
+            Length = length;
+        }
+
+        public int Length { get; }
+    }
+}
+";
+            var comp = CreateCompilation(src, options: TestOptions.DebugDll.WithMetadataImportOptions(MetadataImportOptions.All));
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:354]: > `.WithMetadataImportOptions(MetadataImportOptions.All)`

Just curious - why is this needed in this test?

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs [hunk 23]
```diff
@@ -0,0 +1,6195 @@
...
+                Assert.False(buffer.TryGetInlineArrayElementType().HasType);
+            }
+        }
+
+        [Fact]
+        public void InlineArrayType_15_Interface()
+        {
+            var src = @"
+[System.Runtime.CompilerServices.InlineArray(10)]
+interface Buffer
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:443]: Consider testing also `record`s (e.g., with synthesized fields).

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs [hunk 24]
```diff
@@ -0,0 +1,6195 @@
...
+    {
+        var x = new C();
+        System.Console.Write(M1(x));
+        M2(x);
+        System.Console.Write(' ');
+        System.Console.Write(M1(x));
+    }
+
+    static int M1(C x) => x.F[0];
+    static void M2(C x) => x.F[(dynamic)0] = 111;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:1322]: Consider adding a test indexing with a `dynamic` variable, not constant. Also, do we have tests indexing with `int` variables (not constants)?

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs (hunk 24 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:1322]: > Consider adding a test indexing with a dynamic variable, not constant. 

There is no such thing a s a dynamic constant, unless it has `null` value. Which is not the case here.

> Also, do we have tests indexing with int variables (not constants)?

I do not find this scenario particularly interesting. But will include it into the test plan for completeness.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs (hunk 22 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:354]: > Just curious - why is this needed in this test?

By default, private fields in classes could be dropped on import. I do not want this to affect the test scenario.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs (hunk 23 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InlineArrayTests.cs:443]: > Consider testing also records (e.g., with synthesized fields).

Captured in the test plan

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1123]: > I guess this is fine, but seems that using `HasDecodedData` here would be useful. We would not need two special values (0 and -1) for `InlineArrayLength` (and mapping between them below) - the `if` would just check for `!HasDecodedData`. We could also have `VerifySealed(expected: true);` in getter of `TypeWellKnownAttributeData.InlineArrayLength`.

I guess I am missing how it could help. The ```HasDecodedData``` property simply tells us if an instance of ```TypeWellKnownAttributeData``` has been allocated for this type. It doesn't tell us why and what data are actually stored in in that instance. If the property returns true, that tells us nothing about whether we have seen any specific well known attribute before.

What am I missing?
